### PR TITLE
feat(#77): adds support for b3 single header.

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR adds support for B3 single header.

Ping @basvanbeek @adriancole

### CODE REVIEW

File: propagation/b3/http.go [hunk 1]
```diff
@@ -31,8 +55,13 @@ func ExtractHTTP(r *http.Request) propagation.Extractor {
 			parentSpanIDHeader = r.Header.Get(ParentSpanID)
 			sampledHeader      = r.Header.Get(Sampled)
 			flagsHeader        = r.Header.Get(Flags)
+			singleHeader       = r.Header.Get(Context)
 		)
 
+		if singleHeader != "" {
+			return ParseSingleHeader(singleHeader)
```
Comment (basvanbeek) [inline propagation/b3/http.go]: only return if successful... if error and we have multi header data we should try that as fallback before we give up

File: propagation/b3/http.go [hunk 2]
```diff
@@ -21,6 +21,30 @@ import (
 	"github.com/openzipkin/zipkin-go/propagation"
 )
 
+type Option func(opts *Options)
+
+type Options struct {
```
Comment (jcchavezs) [inline propagation/b3/http.go]: Not sure I should call this `Options` or `InjectOptions`. WDYT @basvanbeek

Comment (basvanbeek) [reply propagation/b3/http.go]: InjectOptions has my preference

File: propagation/b3/http.go [hunk 3]
```diff
@@ -31,8 +55,15 @@ func ExtractHTTP(r *http.Request) propagation.Extractor {
 			parentSpanIDHeader = r.Header.Get(ParentSpanID)
 			sampledHeader      = r.Header.Get(Sampled)
 			flagsHeader        = r.Header.Get(Flags)
+			singleHeader       = r.Header.Get(Context)
 		)
 
+		if singleHeader != "" {
+			if sc, err := ParseSingleHeader(singleHeader); err == nil {
```
Comment (jcchavezs) [inline propagation/b3/http.go]: If we error parsing single header then we forward this concern to the multi header.

Comment (basvanbeek) [reply propagation/b3/http.go]: almost... if ParseHeaders fails too, I'd rather return the error from the single header parser (if single header was available)

Comment (jcchavezs) [reply propagation/b3/http.go]: Makes sense.

Comment (jcchavezs) [reply propagation/b3/http.go]: I also added more tests for checking this fallback too.

File: propagation/b3/http.go [hunk 4]
```diff
@@ -21,52 +21,104 @@ import (
 	"github.com/openzipkin/zipkin-go/propagation"
 )
 
+type InjectOption func(opts *InjectOptions)
+
+type InjectOptions struct {
+	shouldInjectSingle   bool
```
Comment (basvanbeek) [inline propagation/b3/http.go]: How about:
`singleHeader` and `multiHeader`?

File: propagation/b3/http.go [hunk 5]
```diff
@@ -21,52 +21,104 @@ import (
...
+	shouldInjectMultiple bool
+}
+
+// WithInjectSingleAndMultiple allows to include both single and multiple
+// headers in the context injection
+func WithInjectSingleAndMultiple() InjectOption {
```
Comment (basvanbeek) [inline propagation/b3/http.go]: Multiple is ambiguous here... as it is now I find it unclear if it means multiple headers (both single B3 and multi B3) or simply multi-header B3...

I'd rename to `WithSingleAndMultiHeader`

File: propagation/b3/http.go [hunk 6]
```diff
@@ -21,52 +21,104 @@ import (
...
+	return func(opts *InjectOptions) {
+		opts.shouldInjectSingle = true
+		opts.shouldInjectMultiple = true
+	}
+}
+
+// WithInjectSingleOnly allows to include only single header in the context
+// injection
+func WithInjectSingleOnly() InjectOption {
```
Comment (basvanbeek) [inline propagation/b3/http.go]: `WithSingleHeaderOnly`

File: propagation/b3/http.go [hunk 7]
```diff
@@ -21,52 +21,104 @@ import (
...
+		opts.shouldInjectSingle = true
+		opts.shouldInjectMultiple = false
+	}
+}
+
 // ExtractHTTP will extract a span.Context from the HTTP Request if found in
 // B3 header format.
 func ExtractHTTP(r *http.Request) propagation.Extractor {
-	return func() (*model.SpanContext, error) {
+	return func() (sc *model.SpanContext, err error) {
```
Comment (basvanbeek) [inline propagation/b3/http.go]: I think I'd remove the return variable declarations to be more explicit about returns...

File: propagation/b3/http.go [hunk 8]
```diff
@@ -21,52 +21,104 @@ import (
...
 			traceIDHeader      = r.Header.Get(TraceID)
 			spanIDHeader       = r.Header.Get(SpanID)
 			parentSpanIDHeader = r.Header.Get(ParentSpanID)
 			sampledHeader      = r.Header.Get(Sampled)
 			flagsHeader        = r.Header.Get(Flags)
+			singleHeader       = r.Header.Get(Context)
 		)
 
-		return ParseHeaders(
+		var sErr error
```
Comment (basvanbeek) [inline propagation/b3/http.go]: I think this reads better:

```go
var sErr error

if singleHeader != "" {
    sc, sErr := ParseSingleHeader(singleHeader)
    if sErr == nil {
        return sc, nil
     } 
}

sc, mErr := ParseHeaders(....)

if mErr != nil && sErr != nil {
    return nil, sErr
}

return sc, mErr
```

File: propagation/b3/spancontext.go [hunk 9]
```diff
@@ -88,3 +88,113 @@ func ParseHeaders(
 
 	return sc, nil
 }
+
+// ParseSingleHeader takes values found from B3 Single Header and tries to reconstruct a
+// SpanContext.
+func ParseSingleHeader(contextHeader string) (*model.SpanContext, error) {
```
Comment (basvanbeek) [inline propagation/b3/spancontext.go]: I find it hard to read all those positional slice arguments...

I would personally do a split on "-" and have a code flow where the amount of items resulting from this split dictate what kind of b3 single header we got and have all items clearly named for their positional arguments.

e.g.  single item can only mean we expect a sample flag: 0, 1, d
two items means we only have traceID + spanID
three items means we have traceID + spanID + debug
four items means we also have parentSpanID (and sample flag could be empty (meaning undecided))

File: propagation/b3/spancontext_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+		expectedErr error
+	}{
+		{"a", ErrInvalidSampledByte},
+		{"3", ErrInvalidSampledByte},
+		{"000000000000007b", ErrInvalidScope},
+		{"000000000000007b00000000000001c8", ErrInvalidScope},
+		{"000000000000007b00000000000001c8-000000000000007b-", ErrInvalidSampledByte},
+		{"000000000000007b00000000000001c8-000000000000007b-3", ErrInvalidSampledByte},
+		{"000000000000007b00000000000001c8-000000000000007b-00000000000001c8", ErrInvalidParentSpanIDValue},
+		{"000000000000007b00000000000001c8-000000000000007b-1-00000000000001c", ErrInvalidParentSpanIDValue},
```
Comment (basvanbeek) [inline propagation/b3/spancontext_test.go]: we need a test for where we do have a parentspan id but sampled flag is undecided

File: propagation/b3/spancontext_test.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+func TestParseHeaderFails(t *testing.T) {
+	testCases := []struct {
+		header      string
+		expectedErr error
+	}{
+		{"a", ErrInvalidSampledByte},
+		{"3", ErrInvalidSampledByte},
+		{"000000000000007b", ErrInvalidScope},
+		{"000000000000007b00000000000001c8", ErrInvalidScope},
+		{"000000000000007b00000000000001c8-000000000000007b-", ErrInvalidSampledByte},
```
Comment (basvanbeek) [inline propagation/b3/spancontext_test.go]: @adriancole would we want to be lenient here and allow empty sampled item to mean undecided when there is no parentspanid item?

File: propagation/b3/spancontext.go (hunk 9 above)
Comment (basvanbeek) [reply propagation/b3/spancontext.go]: another option could be to define constants for the offsets so by naming it, it is more clear what the payload of the resulting slices is.

**Comment by jcchavezs** [comment]
@basvanbeek so I did some benchmarks about using positioning vs using split and got the following results:
```
$ go test -run BenchmarkParseSingleHeader -bench 10
goos: darwin
goarch: amd64
pkg: github.com/openzipkin/zipkin-go
BenchmarkSpan_100Annotations-4    	  100000	     14234 ns/op
BenchmarkSpan_1000Annotations-4   	   10000	    129253 ns/op
BenchmarkSpan_100Tags-4           	  200000	     12954 ns/op
BenchmarkSpan_1000Tags-4          	   10000	    135022 ns/op
PASS
ok  	github.com/openzipkin/zipkin-go	7.027s

$ go test -run BenchmarkParseSingleHeaderWithSplit -bench 10
goos: darwin
goarch: amd64
pkg: github.com/openzipkin/zipkin-go
BenchmarkSpan_100Annotations-4    	  100000	     14130 ns/op
BenchmarkSpan_1000Annotations-4   	   10000	    129952 ns/op
BenchmarkSpan_100Tags-4           	  200000	     11111 ns/op
BenchmarkSpan_1000Tags-4          	   10000	    132360 ns/op
PASS
ok github.com/openzipkin/zipkin-go 6.625s
```

You can find both methods at https://gist.github.com/jcchavezs/9064b9168cb2b02590c57f677ab5dd4c

**Comment by adriancole** [comment]
let's land this soon if possible. I'd like to cite it on the B3 README cc @jeqo 

https://github.com/apache/incubator-zipkin-b3-propagation/pull/34

**Comment by basvanbeek** [comment]
sometimes it's just better to let it run... I'm not overly happy with the performance loss of the more readable solution so let's start with this and make sure we have some really good tests to avoid regressions and catch issues.

**Comment by jcchavezs** [comment]
Since this parsing is internal. We can switch to something in the middle or
something else at any point (as long as unit tests are strong enough to
catch issues).

tor. 9. mai 2019, 19:49 skrev Bas van Beek <notifications@github.com>:

**Pull request merged**